plutus-core/index-envs/src/Data/RandomAccessList/SkewBinary.hs

```diff
@@ -74,122 +76,94 @@ uncons = \case
-in the GHC Core, allowing GHC to inline the arguments of 'safeIndexOneCont' and transform the whole
+within 'contIndexZero' instead of being pulled out via full-laziness or some other optimization
+pass. This ensures that when 'contIndexZero' gets inlined, the local definitions appear directly
+in the GHC Core, allowing GHC to inline the arguments of 'contIndexZero' and transform the whole
 thing into a beautiful recursive join point full of @Word64#@s, i.e. allocating very little if
 anything at all.
 -}
 
--- See Note [Optimizations of safeIndexOneCont].
-safeIndexOneCont :: forall a b. b -> (a -> b) -> RAList a -> Word64 -> b
-safeIndexOneCont z f = findTree where
+-- See Note [Optimizations of contIndexZero].
+contIndexZero :: forall a b. b -> (a -> b) -> RAList a -> Word64 -> b
+contIndexZero z f = findTree where
     findTree :: RAList a -> Word64 -> b
-    -- See Note [Optimizations of safeIndexOneCont].
+    -- See Note [Optimizations of contIndexZero].
     findTree Nil !_ = z
     findTree (BHead w t ts) i =
-        if i <= w
+        if i < w
         then indexTree w i t
         else findTree ts (i-w)
 
     indexTree :: Word64 -> Word64 -> Tree a -> b
-    -- See Note [Optimizations of safeIndexOneCont].
-    indexTree !w 1 t = case t of
+    -- See Note [Optimizations of contIndexZero].
+    indexTree !w 0 t = case t of
         Node x _ _ -> f x
         Leaf x     -> if w == 1 then f x else z
-    indexTree _ 0 _ = z -- "index zero"
     indexTree _ _ (Leaf _) = z
     indexTree treeSize offset (Node _ t1 t2) =
         let halfSize = unsafeShiftR treeSize 1 -- probably faster than `div w 2`
-            offset' = offset - 1
-        in if offset' <= halfSize
-           then indexTree halfSize offset' t1
-           else indexTree halfSize (offset' - halfSize) t2
-{-# INLINE safeIndexOneCont #-}
+        in if offset <= halfSize
+           then indexTree halfSize (offset - 1) t1
+           else indexTree halfSize (offset - 1 - halfSize) t2
+{-# INLINE contIndexZero #-}
+
+contIndexOne :: forall a b. b -> (a -> b) -> RAList a -> Word64 -> b
+contIndexOne z _ _ 0 = z
+contIndexOne z f t n = contIndexZero z f t (n - 1)
+{-# INLINE contIndexOne #-}
+
+-- 0-based
+unsafeIndexZero :: RAList a -> Word64 -> a
+unsafeIndexZero = contIndexZero (error "out of bounds") id
+{-# INLINE unsafeIndexZero #-}
+
+-- 0-based
+safeIndexZero :: RAList a -> Word64 -> Maybe a
+safeIndexZero = contIndexZero Nothing Just
+{-# INLINE safeIndexZero #-}
+
+-- 1-based
+unsafeIndexOne :: RAList a -> Word64 -> a
+unsafeIndexOne = contIndexOne (error "out of bounds") id
+{-# INLINE unsafeIndexOne #-}
 
 -- 1-based
 safeIndexOne :: RAList a -> Word64 -> Maybe a
-safeIndexOne = safeIndexOneCont Nothing Just
+safeIndexOne = contIndexOne Nothing Just
+{-# INLINE safeIndexOne #-}
 
 instance RAL.RandomAccessList (RAList a) where
     type Element (RAList a) = a
 
-    {-# INLINABLE empty #-}
     empty = Nil
-    {-# INLINABLE cons #-}
+    {-# INLINE empty #-}
```

Comment:
We really want all of that to inline, no point is chickening out and saying `INLINABLE`.